docs/general/cairo_precompiles.md

```diff
@@ -0,0 +1,170 @@
+        self.counter.write(self.counter.read() + 1);
+    }
+
+    #[external(v0)]
+    pub fn set_counter(ref self: ContractState, new_counter: u256) {
+        self.counter.write(new_counter);
+    }
+
+}
+```
+
+Calling this contract from an EVM contract would look like this:
+
+```solidity
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.7.0 <0.9.0;
+
+import "./ICairo.sol";
+
+contract MyContract {
+    /// @dev The Cairo precompile contract's instance.
+
+    /// @dev The cairo contract to call - assuming it's deployed at address 0xabc
+    uint256 constant CAIRO_CONTRACT_ADDRESS = 0xabc;
+
+    /// @dev The cairo function selector to call - `inc()`
+    uint256 constant FUNCTION_SELECTOR_INC =
+    0x03b82f69851fa1625b367ea6c116252a84257da483dcec4d4e4bc270eb5c70a7;
+
+    /// @dev The cairo function selector to call - `set_counter()`
+    uint256 constant FUNCTION_SET_COUNTER = 0x0107cf8c3d109449e1beb4ac1ba726d3673b6f088ae454a9e0f18cb225be4712;
+
+    /// @notice Calls the Cairo contract to increment its internal counter
+    /// @return The return data of the call, serialized as bytes
+    function incrementCairoCounter() public returns (bytes memory) {
+        // `inc` takes no arguments, so data is empty
+        bytes memory data = "";
+
+        return CAIRO_PRECOMPILE_CONTRACT.call_contract(CAIRO_CONTRACT_ADDRESS, FUNCTION_SELECTOR_INC, data);
+    }
+
+    /// @notice Calls the Cairo contract to set its internal counter to an arbitrary value
+    /// @dev The counter value is split into two 128-bit values to match the Cairo contract's expected inputs (u256 is composed of two u128s)
+    /// @param newCounter The new counter value to set
+    /// @return The return data of the call, serialized as bytes
+    function setCairoCounter(uint256 newCounter) public returns (bytes memory) {
+        uint128 newCounterLow = uint128(newCounter);
+        uint128 newCounterHigh = uint128(newCounter >> 128);
+        bytes memory data = abi.encode(newCounterLow, newCounterHigh);
+
+        return CAIRO_PRECOMPILE_CONTRACT.call_contract(CAIRO_CONTRACT_ADDRESS, FUNCTION_SET_COUNTER, data);
+    }
+}
+```
+
+Once deployed, the contract can be called to increment the counter in a Cairo
+contract deployed at address `0xabc`. The deployment address will need to be
+communicated to Kakarot for the precompile to be whitelisted.
+
+Internally, a new logic flow is implemented when processing message calls. If
+the target address is the Cairo precompile address, we check if the code_address
+of the message is whitelisted. If it is, we execute the Cairo contract. If it is
+not, we revert the transaction.
+
+To execute the Cairo contract, we need to convert the calldata in bytes to the
+expected Cairo inputs, serialized to an `Array<felt252>`. This serialization
+process must be done upfront, in the Solidity contract, before calling the
+precompile. Each 256-bit word in the calldata is truncated to 251 bit and packed
```

Comment:
the CALLDATA (`bytes`, ie `bytes1[]`) that we get in the EVM CALL (in Cairo), which is `abi.encode(signature)` is actually an `Array<felt252>` where each `felt < 256`

So at the Cairo level, we need to
- pick the corresponding arguments by slicing this big `felt*`
- pack each 32 bytes word into a (truncated) felt

it makes sense, we sort of need to introduce a x32 coef in the calldata size because each byte is already a felt, so we need 32 felt for 1 felt

Reply:
you can rephrase this `Each 256-bit word in the calldata is truncated to 251 bit and packed` with more context, and it's not 251 but just say native field element (approx 252 bits currently)

Reply:
maybe add an example with the counter `inc(felt)` and call with `data = [0, 0, ..., 0, 1]` to call `inc(1)`

Reply:
done